docs/adr/0007-preventing-version-mismatches.md

```diff
@@ -0,0 +1,233 @@
+- Tell UniFFI users to create a new crate in their workspace with for their bindgen binary.  The code would look
+  something like this:
+
+```
+fn main() {
+    uniffi_bindgen::run_bindgen_app(vec![
+       uniffi_bindgen_language1::clap_command(),
+       uniffi_bindgen_language2::clap_command(),
+       uniffi_bindgen_language3::clap_command(),
+       ...
+    ])
+}
+```
+
+Support for each languages could be behind feature gates, as in `[1]`.  However, libraries could also choose to define a
+preferred set of languages and require that consumers create their own bindgen crates if they want a different set. For
+example, application-services might have a bindgen crate that supports Kotlin and Swift and moz-central could consume
+the application-services components, but use a bindgen crate that supports JS.
+
+Pros:
+  - Doesn't rely on a system-wide install of `uniffi-bindgen`.
+  - Bindgen tool versions are specified in a `Cargo.toml`.
+  - Doesn't require building the main library to generate bindings.
+  - Allows libraries consumers to define their own bindgen crates.
+
+Cons:
+  - Extra boilerplate to create the bindgen crate.
+  - Bindgen tool versions are specified in a different `Cargo.toml` than where the `uniffi` version is specified.
+  - Some libraries don't use workspaces, we would be forcing them to switch to using workspaces.  Note though, that
+    users wouldn't have to drastically change their project layout.  They could keep their existing crate in the same
+    place and make it a [root package](https://doc.rust-lang.org/cargo/reference/workspaces.html#root-package).
+  - If library consumers wants
+
+### **[Option 3] Tool to manage UniFFI installed versions**
+
+We create a tool that manages the bindgen binaries ourselves.
+
+  - Users specify their bindgen tool version in the `uniffi.toml` bindings config
+  - We update `uniffi-bindgen` to run the bindgen tool specified in `uniffi.toml`.  `uniffi-bindgen generate [language]`
+    would:
+    - Lookup the bindgen tool version from `uniffi.toml` for `[language]`
+    - Installs that version of the tool inside the `target` directory ( `cargo install --root target/uniffi --version
+      X.Y.Z uniffi_bindgen_[language]`). We should also store the installed versions and avoid trying to re-install if
+      the version hasn't changed
+    - Runs `target/uniffi/uniffi_bindgen_language` and passes it all additional args from the command line.
+
+
+Pros:
+  - Doesn't rely on a system-wide install of `uniffi-bindgen`.
+  - Doesn't require building the main library to generate bindings.
+  - Doesn't require any additional crates.
+
+Cons:
+  - Requires the most amount of work on our part.
+  - Specifying versions in `uniffi.toml` as well as `Cargo.toml` could be confusing.
+  - If a library consumer wanted to generate bindings for a language that wasn't specified in `uniffi.toml`, then they
+    would need to manage the installed binary themselves.
+
```

Comment:
Trying to think of other ways out of this bind...I wonder what it would look like to declare a uniffi version number in the `.udl` file itself. Syntax would be fiddly as always, but I'm imagining something  like:

```
[UniFFI=1.1]
namespace my_example {
   ...
}
```

This would be clunky and annoying, but it's putting the information in the one place that every component of the system has to look. If you try to generate the scaffolding with an incompatible version of UniFFI, it can detect this and generate a clean helpful error. If you try to generate bindings with an incompatible version of the tool, it can detect this and generate a clean helpful error.

In terms of the goals of the ADR, this would punt entirely on "If possible, developers shouldn't have to worry or know about this issue" because in fact they would have to know about it *more*. But it would let us to a much better job of "If that's not possible, then it should be easy for them to identify and fix the issue". In fact the completely custom error message could direct the user to a page in the manual with tips about how to keep the versions of their tools in sync.

You could treat this version number as the version of UniFFI itself, or as a separate and less-frequently-updated version number that is specifically *the version of UDL* rather than of any particular tool.

Reply:
This is an interesting idea, particularly if we can decouple the crate version from the "ffi version". Ben and I were also chatting about how the existing mechanism seems to have saved us approximately zero times, mainly because of the way app-services ensures the versions match via the intermediate tool - so we were trying to think of a way to kill this entirely in a "safe enough" way - but haven't quite come up with anything yet :)

Reply:
If we decouple the crate version from the FFI version, then something like this seems like a great idea. It simplifies the task of explaining to devs which versions of uniffi are compatible with which versions of the bindgen tools.  I also think the concept of a UDL syntax version would have been nice as we were working through the wrapped/custom types code.  In hindsight it seems like we didn't have a great language to discuss the changes we were making.

Another way to handle this would be to have `uniffi_build` generate some sort of intermediate artifact that includes the version.  In addition to writing `[component].uniffi.rs` to the out dir, it would also serialize the `ComponentInterface` and write it out to the target dir.  The bindgen tool would input that, check the version, then write out its bindings.

I'm not sure if this is better or worse.  Users wouldn't specify the version in the UDL file.  That makes it easier to get started,  but having that version implicit might make compatibility issues harder to understand.

One reason I'm thinking about this is that Mark brought up replacing the UDL file with a macro.  In that world, we might want to do the serialization anyways to avoid having to parse the Rust source tree for each bindings language.

Reply:
> Another way to handle this would be to have uniffi_build generate some sort of intermediate artifact that includes
> the version. 

I like to use WebAssembly Interface Types and wasm-bindgen as a touch-point here. In that case, compiling the Rust code produces both the compiled code, and a binary "interface definition" which can then be processed by downstream tools. It seems like a really nice workable model, and you can imagine the UDL as a kinda of temporary hand-written version of that interface definition.

(In the wasm-bindgen case, the definition is actually generated as an additional custom section of the compiled library, not as a separate artifact, which I think is really neat and makes it less likely for build system shenanigans to accidentally use a mis-matched `.so` and interface face...but may be a bit much mucking around for our purposes here).